Re-factor functions to wrangle wfhz and muac data

tomaszaba opened the pull request:
Ernest, in this PR I addressed issues #26 #27  #28. There were some tasks related to issues #24 #25. For these two I was not quite sure if you did not address them yet. Please let me know. 

## Now, onto the changes made and the rationale behind

1. In `utils.R`: 
  + In `flag_outliers() and `remove_flags()`: 
       - I changed class of the `@param x` from `double` to `numeric` due to the presence of the absolute MUAC values. In principle, if it was just WFHZ or MFAZ, it would be correct to say `double` as they are. For MUAC at this stage, when `flag_muac()` is used/called, MUAC would have been set to millimetres; A MUAC in mm would be an integer. Thus, a middle point for both type of vectors would be `numeric`.

     - I changed the argument unit to `.from` as I believe it makes more sense like that. Maybe `.in` could also make sense? 

    -  In the choices, I changed "crude" to "absolute" to align with the use of absolute MUAC values in the documentation. 
 
+ In `recode_muac()`: 
    - You will see that I in the documentation I changed to class `double` or `numeric` as I want to capture the fact that MUAC can come in centimeters or millimeters. For the latter, I tried to put integer but it fails as checking with either "is(x, "integer") or is.integer(x) fails (FALSE). 
 
Up to now, all functions inside `utils.R` are exported. 

+ In `mw_wrangle_muac()` I did not check if age is numeric or not. This is because: 
   - This would have been checked in `mw_wrangle_age()`
   - I did not feel like it would be necessary because this vector is just used to control flow, what is actually used is age_days from `mw_wrangle_age()`, which is an output that is checked in the tests. 
   - I added an argument `.decimals` so user can control the number of decimal places for the zscores 
   - I did not check the class of sex. Although I indicated that they should either be `character()` or `numeric()`, I think checking for this can be measleading as users can supplie any vector. You may say that this is what is in my docs, which is true, but later on in the docs I mention this specificity about this `character` or `numeric()`. To me, this is what I don’t think I should check for. happy to read your review though. 

## On the general file organisation 
+ I split the functions in two files. This is to address your comment about my files being too long. So I have decided to put one major function into one `.R`. 

Thanks

tomaszaba commented:
I have just made some changes in three files. I realised that my application of NSE was inconsistent; I had some lines of code where I had difused and tidy evaluated some arguments but I was simply not using them. So I fixed this in `mw_wrangle_age`, `mw_wrangle_wfhz` and `mw_wrangle_muac`, while keeping the code neat.

ernestguevarra commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+#' the sex variable is a character vector of values "m" for boys and "f" for girls
+#' and will recode them to 1 and 2 respectively.
+#'
+#' @param age A vector of class `numeric` of child's age in months.
```
why do you have an age parameter when your function doesn't have it in its specification?

tomaszaba replied:
@ernestguevarra , this parameters are used in `mw_wrangle_muac()`. If run `devtools::check()` you will see no warning about this. 
The function was moved to another file, but linked to the same `@rdname`. Running `help(mw_wrangle_muac)` renders everything in one documentation.

ernestguevarra replied:
> @ernestguevarra , this parameter is used in `mw_wrangle_muac()`. When you run `devtools::check()` you do not see warnings about this. You may have had this question for the fact that you don't see the function in this file as I moved into another file, but they are linked to the same `@rdname`. Running `help(mw_wrangle_muac)` renders everything in one documentation.

functions with the same @rdname should be in the same file

tomaszaba replied:
OK. I have created a separate documentation for `mw_wrangle_muac()` and disconnected it from `mw_wrangle_wfhz()`.

ernestguevarra commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+#'
+#' @param age A vector of class `numeric` of child's age in months.
+#'
+#' @param muac A vector of class `numeric` of child's age in months.
+#'
+#' @param .recode_muac Logical. Default is `FALSE`. Set to `TRUE` if MUAC values
+#' should be converted to either centimeters or millimeters.
```
why do you have these documentation for these parameters when the function doesn't have these specified?

tomaszaba replied:
Same as above.

ernestguevarra commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+#' @param .recode_muac Logical. Default is `FALSE`. Set to `TRUE` if MUAC values
+#' should be converted to either centimeters or millimeters.
+#'
+#' @param .to A choice of the unit to which the MUAC values should be converted.
+#' "cm" for centimeters, "mm" for millimeters and "none" to leave as it is.
```
why do you have this parameter documentation when your function doesn't have this in its specification?

tomaszaba replied:
Same as above.

ernestguevarra commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  ## Capture expressions to evaluate later ----
+  recode_sex <- quote(
+    if (.recode_sex) {
+      sex <- ifelse({{ sex }} == "m", 1, 2)
+    } else {{{ sex }}}
+  )
```
this assumes that the only other possible values for sex is "m" and "f". Is that correct? If so, then you either need to state this in the documentation or if you are going to restrict it to just "m" and "f" as alternative, then why not just state that sex should be coded to 1 and 2 before using this function.

Otherwise, you need to allow for other possible codings for male and female and let the user tell you that. I think I showed you code before of how you can do this. If you think this is not important then disregard.

tomaszaba replied:
On this one and some other comments you are now raising I had addressed them, maybe not in a coherent manner, in the first documentation before we started revising it. Then your suggestion was to shorten it on basis of the fact that my users would will know how to deal with this... I will opt to keep this as it is.

ernestguevarra replied:
> On this one and some other comments you are now raising I had addressed them in the first documentation before we started revising it; maybe there were not coherent. Then your suggestion was to shorten it on basis of the fact that my users would will know how to deal with this... I will opt to keep this as it is.

then you need to have checks in your code for coding for sex that is not "m" or "f" and not 1 or 2 and produce an error message that will tell them what is needed. You should describe this in your documentation as well.

ernestguevarra reviewed (commented):
@tomaszaba, see my detailed feedback below. I make suggestions. You decide if you want to follow them or not.

On testing, I did `devtools::test_coverage_active_file()` when the test file for wrangle_muac is open and I note 5 lines of code  note covered by testing. Please add test that cover these five lines of code.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+
+#'
+#'
+#' Identify and flag outliers
+#'
+#' @description
+#' Outliers are extreme values that deviate remarkably from the survey mean, making
+#' them unlikely to be accurate measurements. This function identifies and signals
+#' them based on a criterion set for the WFHZ, the MFAZ and for the absolute MUAC
+#' values. For the absolute MUAC values, outliers are identified and flagged using
+#' a different approach as described below.
```
> Important to make it clear that this flagging approach is based on SMART methodology. I know you put that in the @details but I would rather that be up here to make it clear that this is the SMART approach to flagging outliers. Make the title clear and specific that this is for flagging outlier z-score values for weight-for-height and for muac-for-age z-score. However, is it correct to say that this can be used for the other z-scores as well (hfaz, wfaz)? I don't remember this well from SMART but if that is true, then I thin you should point out the general use for this function.

Yes, SMART flags are the same for either WFHZ, WFAZ or HFAZ. 

> Regarding the language in the description, what do you mean by remarkably? and what do you mean by survey mean? I think it needs to be clear that this flagging approach is specific for anthropometric surveys where a survey mean can be calculated and then briefly state what the survey mean means.

I think trying to explain too much here will make the description long. Also, I provide links to SMART so users can go and get more details. 

> I am not sure what the term signals here mean. It doesn't seem coherent. I will use indicate instead or just say identifies.

OK.

> What do you mean by absolute MUAC? I know what you mean about it but I am pretty sure that is not what the appropriate terminology is for stating raw MUAC. I think raw or unadjusted or non-standardised MUAC makes more sense. Absolute has a very different meaning in different contexts.

OK, so we change to raw. We came from crude to absolute and now to raw. 

> Also, since you use the term flag, flagging, flagged here and in the docs, please give a quick/short explanation of what a flag is and what a flagged record is.

Wouldn't this make the description long? Your initial review was to avoid having a long description and you really stressed on that. Isn't this going to override that principle?

ernestguevarra replied:
You write long description that doesn't explain the more important things that users won't know. Flagging, flag is not a universal concept. You understand it because you know of it and you use it. Other users won't. I reviewed a package that used another terminology for z-score calculation flagging for CDC growth charts, and one of the comments for that was their use of another alternative term for flagging. These are not words known to everyone.

ernestguevarra replied:
> I think trying to explain too much here will make the description long. Also, I provide links to SMART so users can go and get more details.

this is not about length, you can explain concepts in a phrase or a sentence. Not in paragraphs. Remarkably different doesn't mean anything to me. I would just say different. then i will say this approach is what is used by SMART methodology. then you have the source.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+#' @param x A vector of class `numeric` of WFHZ or MFAZ or absolute MUAC values.
+#' The latter should be in millimeters.
+#'
+#' @param .from A choice between `zscores` and `absolute` for where outliers should be
```
Ok. Will change to raw as you suggested earlier.

tomaszaba commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+#' be 1 or "m" for boy and 2 or "f" for girl. The variable name must be sex,
+#' otherwise it will not work.
+#'
+#' @param .recode_sex Logical. Default is `FALSE`. Setting to `TRUE` assumes that
+#' the sex variable is a character vector of values "m" for boys and "f" for girls
+#' and will recode them to 1 and 2 respectively.
```
OK.

tomaszaba replied:
@ernestguevarra , I think this may be confusing. The second part where you say "Otherwise, set to FALSE (default)". I think this doesn't link to the fact that this only works on vector of "m" and "f".

ernestguevarra replied:
> @ernestguevarra , I think this may be confusing. The second part where you say "Otherwise, set to FALSE (default)". I think this doesn't link to the fact that this only works on vector of "m" and "f".

That second part means that if values for sex is indeed 1 and 2, then this should be set to default. I am not sure that documentation that says "assumes" works well. this relates to my feedback below on sex. Again, I find your documentation unnecessarily long.

tomaszaba replied:
OK. Done.

tomaszaba commented:
> @tomaszaba, see my detailed feedback below. I make suggestions. You decide if you want to follow them or not.
> 
> On testing, I did `devtools::test_coverage_active_file()` when the test file for wrangle_muac is open and I note 5 lines of code note covered by testing. Please add test that cover these five lines of code.

@ernestguevarra, thanks for the review. I'll reflect on these and take the appropriate action.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+  ## Check if unit's arguments match ----
+  .to <- match.arg(.to)
+
+  # ## Check if the class of vector "x" is "numeric" or "double" ----
+  # if (!is.numeric(x) | !is.double(x)) {
+  #   stop(
+  #     "MUAC should be of class 'numeric' or 'double'. Please try again."
+  #   )
+  # }
```
Thanks. I did had not noticed this.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+#' values. For the absolute MUAC values, outliers are identified and flagged using
+#' a different approach as described below.
+#'
+#' @param x A vector of class `numeric` of WFHZ or MFAZ or absolute MUAC values.
+#' The latter should be in millimeters.
```
> Can you add a check in the function that will make a sensible guess whether the MUAC provided is in cms or in millimetres? I think this should be possible, no?

@ernestguevarra , I am very keen to see how this is done. Could you share an example? 

> For this function, compared to your other functions that work with vectors also, this function will provide a wrong answer if x is a vector of 1 only. This is because the mean zscore will just be based on one value and the flagging will not work. This is particularly going to be an issue for users who are more familiar with the WHO flagging criteria which is not based on the survey mean but just based on cut-offs. For flagging based on cut-offs, vectors of 1 can be assessed appropriately. But for this, a vector of one will not cause an error but will provide a meaningless answer. Do you agree?

Yes, I agree. But, do you think this would create that confusion? What I see is that later on, users may request to add an option to identify flags based on WHO flagging criteria. This would, in its turn, imply re-factoring the prevalence calculators to. This is available in ENA, but almost not used. I suggest we wait and see reactions/suggestion and then we re-factor it. 

> So, if you agree, I think we need to have a check here that the x is indeed a vector of more than 1 value and is of a reasonable sample size (length) for this function to return an appropriate answer.

Ok. I will give it a try.

tomaszaba replied:
@ernestguevarra, on checking if MUAC is in 'cm' or 'mm', I approached this as follows: 

```r
recode_muac <- function(x, .to = c("cm", "mm")) {
  ## Check if unit's arguments match ----
  .to <- match.arg(.to)

  ## Check if the class of vector "x" is "numeric" or "double" ----
  if (!is.numeric(x) | !is.double(x)) {
    stop(
      "MUAC should be of class 'numeric' or 'double'. Please try again."
    )
  }

  ## Recode muac conditionally ----
  switch(.to,
    ### Recode to centimeters ----
    "cm" = {
      #### Check if MUAC values are in "mm" ----
      if(any(grepl("\\.", as.character(x)))) {
        stop("The MUAC values supplied are not in millimeters. Please try again.")
      }
      #### Convert MUAC to cm ----
      z <- (x / 10)
      z
    },

    ### Recode to millimeters ----
    "mm" = {
      #### Check if MUAC values are in "cm" ----
      if(all(!grepl("\\.", as.character(x)))) {
        stop("The MUAC values supplied are not in centimeter. Please try again.")
      }
      #### Convert MUAC to mm ----
      z <- (x * 10)
      z
    }
  )
}
```

What do you think about this?

tomaszaba commented on R/wrangle_muac.R:
Indeed. I am not huge fun of package::function(), so I will keep it clean with `usethis::use_import_from()`.

tomaszaba commented on R/wrangle_wfhz.R:
```diff
@@ -0,0 +1,146 @@
+#'
+#' @param df A dataset of class `data.frame` to wrangle data from.
+#'
+#' @param sex A `numeric` or `character` vector of child's sex. Code values should
+#' be 1 or "m" for boy and 2 or "f" for girl. The variable name must be sex,
+#' otherwise it will not work.
```
I agree. I had this in my previous documentation and you didn't like it.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+  ## Identify and flag outliers from zscores ----
+  if (.from == "zscores") {
+    mean_zscore <- mean(x, na.rm = TRUE)
+    flags <- ifelse((x < (mean_zscore - 3) | x > (mean_zscore + 3)), 1, 0)
```
Removed.

tomaszaba commented on R/utils.R:
```diff
@@ -53,3 +53,166 @@ get_age_months <- function(dos, dob) {
+#'
+#' @details
+#' The flagging criterion used for the WFHZ and the MFAZ is as in
+#' [SMART plausibility check](https://smartmethodology.org/). A fixed flagging
+#' criterion is used for the absolute MUAC values. This is as recommended by
+#' [Bilukha, O., & Kianian, B. (2023).](https://doi.org/10.1111/mcn.13478)
```
Done. It looks nicer using the `@reference` tag. Thanks for this tip.

tomaszaba commented:
Hi @ernestguevarra, 
I've just pushed some commits on changes made in the code re-factoring and improving the clarity and the comprehensiveness of the function documentation. Please see below an overview of the changes made: 

1. **Function documentation:**
   - I indicated the function's behaviour or reaction when a wrong input vector is provided opposite to the expected. I did this in all relevant `@params` of `.R` files we have re-factored thus far. 
   - I added a note that gives a context about the flags or flagged record. I hope this is fine.
   - Removed the links and used `@reference` instead. It looks nice, neat and tidy. Thanks for your suggestion.

2. **Re-codes:**
   - Re-factored the function checks to ensure that only expected inputs are supplied. I have to say, working on this task was quite satisfying. It reminded me of common programmes that I use that when I supply a wrong input they error and display a message. That felt like really programming. At the beginning I lacked the idea on how to address this task. Then I checked some of your repos and [this resource](https://adv-r.hadley.nz/conditions.html#signalling-conditions). I also realised that I still have to workout a lot on the power of string manipulation and regular expressions. As usual, thanks for sharing your wisdom with me. 
   - On `recode-muac()` I decided to broaden the checks to include `integers` too. This is to avoid an unnecessary issues where `R` would deny a MUAC vector that would be of class integer. There was no harm in doing it, so I did it. 
 
3. **Test coverage**:
   - `wrangle_age.R`: 100%
   - `utils.R`: 100%
   - `wrangle_wfhz.R`: 100%
   - `wrangle_muac.R`: 97.06% In this file, there is a line of code that I couldn't figure out how to test it. I thought it would have been addressed in the "mother" function `recode_muac()`. 

Looking forward to your feedback.

ernestguevarra reviewed (approved):
@tomaszaba please merge to `dev` or `main` whichever one you want.
